# Load libraries

In [191]:
import numpy as np 
import pandas as pd 

# Load dataset

In [15]:
df = pd.read_csv('Downloads/store_sku_ba_dataset.csv')

In [17]:
df.shape

(1000, 5)

# Dataset columns info

In [21]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Store ID      1000 non-null   object 
 1   SKU ID        1000 non-null   object 
 2   Total Visits  1000 non-null   int64  
 3   Transactions  1000 non-null   int64  
 4   Revenue       1000 non-null   float64
dtypes: float64(1), int64(2), object(2)
memory usage: 39.2+ KB


In [23]:
df.head(5)

Store ID SKU ID  Total Visits  Transactions   Revenue
0  Store_1  SKU_1           606           573  74688.67
1  Store_1  SKU_2           695            27   1390.60
2  Store_1  SKU_3           389           285  50679.63
3  Store_1  SKU_4           689           505  16764.34
4  Store_1  SKU_5           522           230  41928.40

# Finding conversion rates

# 1) Product level analysis

In [29]:
# product level analysis
df['sku_conversion_rate']=df['Transactions']/df['Total Visits']

In [76]:
# top 5 Product level Conversion rates
df.sort_values(by='sku_conversion_rate', ascending=False).head(5)

Store ID SKU ID  Total Visits  Transactions    Revenue  \
401  Store_21  SKU_2           538           537    9392.72   
540  Store_28  SKU_1           734           732  117151.27   
22    Store_2  SKU_3           886           883  173818.90   
724  Store_37  SKU_5           647           644    7789.79   
281  Store_15  SKU_2           194           193   21862.08   

     sku_conversion_rate  
401             0.998141  
540             0.997275  
22              0.996614  
724             0.995363  
281             0.994845

 SKU_1,SKU_2,SKU_3 ,etc  have max conversion rates (>99%) - they are the top performing products

In [164]:
# Top averge conversion rates per SKU
sku_summary = df.groupby('SKU ID').agg({
    'Total Visits': 'sum',
    'Transactions': 'sum',
    'Revenue': 'sum'
}).reset_index()

sku_summary['overall_conversion_rate'] = sku_summary['Transactions'] / sku_summary['Total Visits']

# Sort by overall conversion rate
sku_summary.sort_values(by='overall_conversion_rate', ascending=False).head(5)

SKU ID  Total Visits  Transactions     Revenue  overall_conversion_rate
2   SKU_11         27856         16323  1665863.73                 0.585978
17   SKU_7         26103         14654  1674024.12                 0.561391
7   SKU_16         23574         12906  1234789.51                 0.547468
14   SKU_4         24782         13505  1370489.58                 0.544952
19   SKU_9         26715         14324  1284551.61                 0.536178

sku_11 has the top mean conversion rate across all stores

In [78]:
df.sort_values(by='sku_conversion_rate', ascending=True).head(10)

Store ID  SKU ID  Total Visits  Transactions  Revenue  \
164   Store_9   SKU_5           464             0     0.00   
189  Store_10  SKU_10           927             0     0.00   
388  Store_20   SKU_9            66             0     0.00   
578  Store_29  SKU_19           846             0     0.00   
311  Store_16  SKU_12           874             0     0.00   
154   Store_8  SKU_15           410             0     0.00   
631  Store_32  SKU_12           707             1    61.80   
33    Store_2  SKU_14           621             1   199.29   
82    Store_5   SKU_3           669             4   387.52   
695  Store_35  SKU_16           897             6   407.33   

     sku_conversion_rate  
164             0.000000  
189             0.000000  
388             0.000000  
578             0.000000  
311             0.000000  
154             0.000000  
631             0.001414  
33              0.001610  
82              0.005979  
695             0.006689

 Products having  SKU ID 5,10,9,19,12,15 have a least conversion rate  of 0 

# 2) Store level analysis

In [35]:
# for store level analysis
store_level_df = df.groupby('Store ID')[['Total Visits', 'Transactions']].sum()

In [37]:
store_level_df.head()

Total Visits  Transactions
Store ID                            
Store_1          11296          6054
Store_10         12173          6311
Store_11          9362          4526
Store_12          8698          3310
Store_13         11846          5155

In [41]:
# finding conversion rates in different stores
store_level_df['Store_Conversion_Rate'] = store_level_df['Transactions'] / store_level_df['Total Visits']

In [43]:
store_level_df.head()

Total Visits  Transactions  Store_Conversion_Rate
Store ID                                                   
Store_1          11296          6054               0.535942
Store_10         12173          6311               0.518442
Store_11          9362          4526               0.483444
Store_12          8698          3310               0.380547
Store_13         11846          5155               0.435168

In [63]:
# Top 5 stores by conversion rate
store_level_df.sort_values(by='Store_Conversion_Rate', ascending=False).head(5)

Total Visits  Transactions  Store_Conversion_Rate
Store ID                                                   
Store_39          9213          6124               0.664713
Store_49         11032          7129               0.646211
Store_44          8482          5356               0.631455
Store_37          9458          5937               0.627723
Store_47          9753          6081               0.623500

 Store 39 has max conversion rate of 66.4%

In [65]:
# Bottom 5 stores by conversion rate
store_level_df.sort_values(by='Store_Conversion_Rate', ascending=True).head(5)

Total Visits  Transactions  Store_Conversion_Rate
Store ID                                                   
Store_15          8922          3090               0.346335
Store_14         12804          4674               0.365042
Store_36         10149          3819               0.376293
Store_30         10659          4043               0.379304
Store_12          8698          3310               0.380547

 Store 15 has min conversion rate of 34.63%

# Stores having Top mean conversion rates

In [145]:
# Group by Store ID and get average SKU conversion rate per store
avg_conversion_per_store = df.groupby('Store ID')['sku_conversion_rate'].mean().sort_values(ascending=False)

# Check top stores
top_avg_conversion_stores = avg_conversion_per_store.head(5)
display(top_avg_conversion_stores)

Store ID
Store_47    0.658466
Store_49    0.611291
Store_39    0.605598
Store_37    0.605362
Store_43    0.583027
Name: sku_conversion_rate, dtype: float64

store_47 has highest mean conversion rate of 65.8%

In [94]:
# Look at SKU conversion rates within one top store
df[df['Store ID'] == top_avg_conversion_stores.index[0]][['SKU ID', 'sku_conversion_rate']].sort_values(by='sku_conversion_rate', ascending=False)

SKU ID  sku_conversion_rate
930  SKU_11             0.992727
922   SKU_3             0.989583
937  SKU_18             0.967213
929  SKU_10             0.904448
933  SKU_14             0.851804
921   SKU_2             0.830957
934  SKU_15             0.825758
926   SKU_7             0.816406
932  SKU_13             0.815152
925   SKU_6             0.800905
924   SKU_5             0.652916
939  SKU_20             0.650206
935  SKU_16             0.576842
931  SKU_12             0.565363
938  SKU_19             0.474880
927   SKU_8             0.359712
923   SKU_4             0.320442
936  SKU_17             0.315315
920   SKU_1             0.275762
928   SKU_9             0.182927

sku_11 is the top product in the top store

In [98]:
# Group by SKU ID, get number of stores and average conversion
sku_performance = df.groupby('SKU ID').agg({
    'Store ID': 'nunique',
    'sku_conversion_rate': 'mean'
}).rename(columns={'Store ID': 'Num_Stores'})

# Filter for SKUs sold in many stores and high conversion
top_skus_across_stores = sku_performance[sku_performance['Num_Stores'] > 3].sort_values(by='sku_conversion_rate', ascending=False).head(5)

In [100]:
print("Top 5 SKUs performing well across multiple stores:")
display(top_skus_across_stores)

Top 5 SKUs performing well across multiple stores:


Num_Stores  sku_conversion_rate
SKU ID                                 
SKU_11          50             0.576090
SKU_7           50             0.565181
SKU_9           50             0.533463
SKU_1           50             0.532657
SKU_16          50             0.528648

# Conversions v/s visits

In [180]:
#  Sort by Total Visits (descending)
top_visited_stores = store_level_df.sort_values(by='Total Visits', ascending=False).head(25)

#  Sort those top visited stores by conversion rate (ascending)
high_visit_low_conv = top_visited_stores.sort_values(by='Store_Conversion_Rate')

#  Show the bottom 5 conversion rates among top 25 visited stores
high_visit_low_conv.head(5)

Total Visits  Transactions  Store_Conversion_Rate
Store ID                                                   
Store_14         12804          4674               0.365042
Store_30         10659          4043               0.379304
Store_19         10678          4627               0.433321
Store_13         11846          5155               0.435168
Store_42         11239          5196               0.462319

Store_14 has the least conversion rate but highest visits

# Low efficiency stores

In [126]:
low_efficiency_stores = store_level_df[
    (store_level_df['Total Visits'] > 1000) &
    (store_level_df['Store_Conversion_Rate'] < 0.4)
]
display(low_efficiency_stores)

Total Visits  Transactions  Store_Conversion_Rate
Store ID                                                   
Store_12          8698          3310               0.380547
Store_14         12804          4674               0.365042
Store_15          8922          3090               0.346335
Store_23          9411          3759               0.399426
Store_30         10659          4043               0.379304
Store_36         10149          3819               0.376293

# Correlation b/w Revenue and sku_conversion_rate

In [132]:
# Check correlation
df[['Revenue', 'sku_conversion_rate']].corr()

Revenue  sku_conversion_rate
Revenue              1.000000             0.568857
sku_conversion_rate  0.568857             1.000000

As correlation value b/w  revenue and sku_conversion rate is 0.568857 (+ve) , SKU conversion rate increases, revenue increases as well.

In [134]:
# High revenue, low conversion SKUs
rev_vs_conv = df[(df['Revenue'] > df['Revenue'].quantile(0.75)) & (df['sku_conversion_rate'] < df['sku_conversion_rate'].quantile(0.25))]

In [138]:
display(rev_vs_conv[['SKU ID', 'Revenue', 'sku_conversion_rate']].sort_values(by='Revenue', ascending=False))

SKU ID   Revenue  sku_conversion_rate
803  SKU_4  41853.74              0.24055

 Above is showing skus having top 25% revenue but but bottom 25% conversion rate